src/components/Widget/inputs.tsx

```diff
@@ -71,20 +71,30 @@ export function useSyncWidgetInputs({
   }, [])
 
   const onTokenSelect = useCallback(
-    (token: Currency) => {
+    (selectingToken: Currency) => {
       if (selectingField === undefined) return
-      setType(toTradeType(selectingField))
 
       const otherField = invertField(selectingField)
-      let otherToken = tokens[otherField]
-      otherToken = otherToken?.equals(token) ? tokens[selectingField] : otherToken
-      const update = {
-        [selectingField]: token,
-        [otherField]: otherToken,
+      const isFlip = tokens[otherField]?.equals(selectingToken)
+      const update: SwapTokens = {
+        [selectingField]: selectingToken,
+        [otherField]: isFlip ? tokens[selectingField] : tokens[otherField],
         default: tokens.default,
       }
+
+      setType((type) => {
+        // If flipping the tokens, also flip the type/amount.
+        if (isFlip) return invertTradeType(type)
+
+        // Setting a new token should clear its amount, if it is set.
+        const activeField = type === TradeType.EXACT_INPUT ? Field.INPUT : Field.OUTPUT
+        if (selectingField === activeField) setAmount(() => EMPTY_AMOUNT)
+
+        return type
```

Comment:
nit: add curlies on the return statement--took me way too long to notice the early return on inversion and thought there was a bug here

Reply:
I think we just need to add this as a lint rule to prevent people (aka _me, the biggest offender_) from doing it.

Reply:
That'd also let us fix it programmatically, instead of wasting code review effort on it.